File: union_find.rb
# Union Find
# http://www.prefield.com/algorithm/container/union_find.html
class UnionFind
  def initialize(n)
    @data = Array.new(n, -1)
  end

  def union_set(a, b)
    a = root(a)
    b = root(b)
    if a != b
      a, b = b, a if @data[b] < @data[a]
      @data[a] += @data[b]
      @data[b] = a
    end
    return a != b
  end

  def find_set(a, b)
    root(a) == root(b)
  end

  def root(a)
    return a if @data[a] < 0
    @data[a] = root(@data[a])
  end
end

File: cpp/union_find.cpp
struct UnionFind {
  vector<int> d;
  UnionFind(int n = 0): d(n, -1) {}

  bool unite(int x, int y) {
    x = root(x);
    y = root(y);

    if (x != y) {
      if (d[x] > d[y]) swap(x,y);
      d[x] += d[y];
      d[y] = x;
    }
    return x != y;
  }

  bool same(int x, int y) {
    return root(x) == root(y);
  }

  int size(int x) {
    return -d[root(x)];
  }

  int root(int x) {
    if (d[x] < 0) return x;
    return d[x] = root(d[x]);
  }
};
